content/blog/hackathon-2022-wazo-matrix-bridge.md

```diff
@@ -0,0 +1,60 @@
+
+In order to create the bridge we decided to use the [Mautrix](https://github.com/mautrix/python) Library since it is very feature complete, written in Python, and they have many well established bridges already. We spent a lot of time looking through the library and looking at the other implementations to figure out how we might do something similar for Wazo. Then we created the base structure of the project and the models needed to store information about the users, puppets and groups in order to create a mapping between the two systems.
+
+## Step 1: Understanding, planning and the database
+
+Our first step, but to figure out how Mautrix works. Their bridges are very feature rich, but we were unable to find much documentation for the actual library, so we needed to spend a lot of time reading the library code and looking at other bridges to figure out how they worked. Also, lots of testing was needed. Then we needed to make the Database models for the different items that need to be saved in the bridge, such as Wazo-Matrix mappings for users and groups. Typically, we use PostgreSQL, but for our test we simply used sqlite because it required less configuration and it would be easier to copy the database around for our tests.
+
+## Step 2: Bridge Wazo to Matrix
+
+In order to test something, we needed at least one direction to work. So we started with the Wazo side. We needed to know as soon as a message was sent to our user so we could create the groups/users and forward the message to Matrix. To do this, we created a small HTTP server using aiohttp and set up a webhook with Webhookd. The webhook listens for an new message event in chatd and then posts the message information to a URL that our bridge exposes. When the message content is received the bridge checks to whom the message is being sent. If the users don't already in the bridge and Matrix, then it must create them and the room. The user of Matrix receives and invite to the new room which contains puppeted users. Once the room and users already exist, the bridge simply forwards the new messages to the correct group in Matrix.
+
+## Step 3: Different message formats
+
+Parallel to the bridge, we were also working on supporting new types of content in chatd such as Markdown, emoji, files and reactions. ([You can read about that here](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo)). The Matrix protocol already supports these things as does the client we were using (Element). In order to support Markdown, we simply used a python library python-markdown to render the text into HTML and send that as the `formatted_body` to Matrix. This allowed for rich text content such as lists, tables, bold, italic, strike-through, embedded images and more. As for emoji, since they are simply unicode characters, no special logic was needed. However, sadly, we did not have time to add reactions or file handling to the bridge during the event.
```

Comment:
```suggestion
Parallel to the bridge, we were also working on supporting new types of content in chatd such as Markdown, emoji, files and reactions. ([You can read about that here](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo)). The Matrix protocol already supports these things as does the client we were using (Element). In order to support Markdown, we simply used a python library python-markdown to render the text into HTML and send that as the `formatted_body` to Matrix. This allowed for rich text content such as lists, tables, bold, italic, strike-through, embedded images and more. As for emojis, since they are simply Unicode characters, no special logic was needed. However, sadly, we did not have time to add reactions or file handling to the bridge during the event.
```

Reply:
Wow, I'm embarrassed you found so many typos. Thanks! I merged all the others, but I'm pretty sure Unicode is supposed to be capitalised since it is a proper noun and emoji is a Japanese loan word and as such shouldn't really be pluralised either.